File: src/main/java/com/plugin/frege/resolve/FregeFunctionNameReference.java
package com.plugin.frege.resolve;

import com.intellij.psi.PsiElement;
import com.plugin.frege.psi.FregeFunctionName;
import org.jetbrains.annotations.NotNull;

import java.util.List;

import static com.plugin.frege.psi.impl.FregePsiUtilImpl.findElementsWithinScope;
import static com.plugin.frege.psi.impl.FregePsiUtilImpl.getByTypePredicateCheckingText;

public class FregeFunctionNameReference extends FregeReferenceBase {
    public FregeFunctionNameReference(@NotNull PsiElement element) {
        super(element, element.getTextRange());
    }

    // TODO improve
    @Override
    protected List<PsiElement> resolveInner(boolean incompleteCode) {
        return findElementsWithinScope(element,
                getByTypePredicateCheckingText(FregeFunctionName.class, element, incompleteCode));
    }
}

File: src/main/kotlin/com/plugin/frege/resolve/FregeFunctionNameReference.kt
package com.plugin.frege.resolve

import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.plugin.frege.psi.FregeFunctionName
import com.plugin.frege.psi.impl.FregePsiUtilImpl

class FregeFunctionNameReference(element: PsiElement) : FregeReferenceBase(element, TextRange(0, element.textLength)) {
    // TODO improve
    override fun resolveInner(incompleteCode: Boolean): List<PsiElement> {
        val firstBinding = FregePsiUtilImpl.findElementsWithinScope(
            psiElement,
            FregePsiUtilImpl.getByTypePredicateCheckingText(FregeFunctionName::class, psiElement, incompleteCode)
        ).minByOrNull { it.textOffset }

        return if (firstBinding != null) listOf(firstBinding) else emptyList()
    }
}
